core/network/impl/gossiper_broadcast.cpp

```diff
@@ -17,7 +17,75 @@ namespace kagome::network {
   void GossiperBroadcast::reserveStream(
       const libp2p::peer::PeerInfo &peer_info,
       std::shared_ptr<libp2p::connection::Stream> stream) {
-    streams_.emplace(peer_info, std::move(stream));
+    reserved_streams_.emplace(peer_info, std::move(stream));
+  }
+
+  void GossiperBroadcast::addStream(
+      std::shared_ptr<libp2p::connection::Stream> stream) {
+    auto peer_id_res = stream->remotePeerId();
+    if (not peer_id_res.has_value()) {
+      logger_->error("Can't get peer_id: {}", peer_id_res.error().message());
+      return stream->reset();
+    }
+    auto &peer_id = peer_id_res.value();
+
+    auto syncing_stream_it = syncing_streams_.find(peer_id);
+    if (syncing_stream_it != syncing_streams_.end()) {
+      if (syncing_stream_it->second != stream) {
+        syncing_stream_it->second->reset();
+        syncing_stream_it->second.swap(stream);
+        logger_->debug("Syncing stream (peer_id={}) was replaced",
+                       peer_id.toHex());
+      }
+      return;
+    }
+
+    logger_->debug("Syncing stream (peer_id={}) was emplaced", peer_id.toHex());
+  }
+
+  uint32_t GossiperBroadcast::getActiveStreamNumber() {
+    struct HashtableHelper {
+      using type = std::reference_wrapper<const libp2p::peer::PeerId>;
+      size_t operator()(const type &item) const noexcept {
+        return std::hash<libp2p::peer::PeerId>()(item.get());
+      }
+      bool operator()(const type &lhs, const type &rhs) const noexcept {
+        return lhs.get() == rhs.get();
+      }
+    };
+
+    std::unordered_set<std::reference_wrapper<const libp2p::peer::PeerId>,
+                       HashtableHelper,
+                       HashtableHelper>
+        unique_peer_ids;
+
```

Comment:
If I'm wright ```syncing_streams_``` and ```reserved_streams_``` should not have cross-subset of streams. Than we don't need any set here. Just several assertions in insertion places and a counter here.

Reply:
It is legacy and will be refactored. Special issue was made: #546